faiss_vector_cache.go

```diff
@@ -0,0 +1,195 @@
+//  Copyright (c) 2024 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build vectors
+// +build vectors
+
+package zap
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	faiss "github.com/blevesearch/go-faiss"
+)
+
+type vecIndexCache struct {
+	closeCh chan struct{}
+	m       sync.RWMutex
+	cache   map[uint16]*cacheEntry
+}
+
+type ewma struct {
+	alpha  float64
+	avg    float64
+	sample uint64
+}
+
+type cacheEntry struct {
+	tracker *ewma
+
+	m     sync.RWMutex
+	index *faiss.IndexImpl
+}
+
+func newVectorIndexCache() *vecIndexCache {
+	return &vecIndexCache{
+		cache:   make(map[uint16]*cacheEntry),
+		closeCh: make(chan struct{}),
+	}
+}
+
+func (vc *vecIndexCache) Clear() {
+	vc.m.Lock()
+	close(vc.closeCh)
+	vc.cache = nil
+	vc.m.Unlock()
+}
+
+func (vc *vecIndexCache) loadVectorIndex(fieldID uint16,
+	indexBytes []byte) (vecIndex *faiss.IndexImpl, err error) {
+	cachedIndex, present := vc.isIndexCached(fieldID)
+	if present {
+		vecIndex = cachedIndex
```

Comment:
Isn't there a possibility (in theory) that this vecIndex could be closed by the time the reader gets around to actually using it. This is related to the comment I added below in line 157.

Reply:
yeah likith had been working on something along the same lines - if a query (on one thread) is executing the search i.e its at faissIndex.SearchWithoutIDs() and the monitor thread (theoretically) ends up closing the index because the avg dropped below during that execution we might end up in bad situations. likith's last patch addresses this by tracking the live refs to the index which will be decremented in Close() API.

Reply:
so the namings in the latest patch is that 

hit(ewma.sample) -> this is used to track the per second hits to the cache entries - helps in computing the average traffic for the field (which is tracked in ewma.avg field) until now. incHit() is the API which increments this value.

refs -> this is used to track the live references to the cache entry, such that while we do a refresh and we see that the avg is below a threshold we close/cleanup only if the live refs to the cache entry is 0. addRef() and decRef() are the corresponding func calls

Reply:
That's good info - should be added as commentary where declared in the code.